drivers/counter/counter_cmos.c

```diff
@@ -0,0 +1,207 @@
+#define STATUS_B_BIN	0x01	/* binary (vs BCD) mode */
+#define HOUR_PM		0x80	/* high bit of hour set = PM */
+
+/*
+ * Read a value from the CMOS. Because of the address latch,
+ * we have to spinlock to make the access atomic.
+ */
+
+static u8_t read_register(u8_t addr)
+{
+	static struct k_spinlock lock;
+	k_spinlock_key_t k;
+	u8_t val;
+
+	k = k_spin_lock(&lock);
+	sys_out8(addr, X86_CMOS_ADDR);
+	val = sys_in8(X86_CMOS_DATA);
+	k_spin_unlock(&lock, k);
+
+	return val;
+}
+
+/* Populate 'state' with current RTC state. */
+
+void read_state(struct state *state)
+{
+	int i;
+	u8_t *p;
+
+	p = (u8_t *) state;
+	for (i = 0; i < sizeof(*state); ++i) {
+		*p++ = read_register(addrs[i]);
+	}
+}
+
+/* Convert 8-bit (2-digit) BCD to binary equivalent. */
+
+static inline u8_t decode_bcd(u8_t val)
+{
+	return (((val >> 4) & 0x0F) * 10) + (val & 0x0F);
+}
+
+/*
+ * Hinnant's algorithm to calculate the number of days offset from the epoch.
+ */
+
+static u32_t hinnant(int y, int m, int d)
+{
+	unsigned yoe;
+	unsigned doy;
+	unsigned doe;
+	int era;
+
+	y -= (m <= 2);
+	era = ((y >= 0) ? y : (y - 399)) / 400;
+	yoe = y - era * 400;
+	doy = (153 * (m + ((m > 2) ? -3 : 9)) + 2)/5 + d - 1;
+	doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
+
+	return era * 146097 + ((int) doe) - 719468;
+}
+
+/*
+ * Returns the Unix epoch time (assuming UTC) read from the CMOS RTC.
+ * This function is long, but linear and easy to follow.
+ */
+
+u32_t read(struct device *dev)
+{
+	struct state state, state2;
+	u64_t *pun = (u64_t *) &state;
+	u64_t *pun2 = (u64_t *) &state2;
+	bool pm;
+	u32_t epoch;
+
+	ARG_UNUSED(dev);
+
+	/*
+	 * Read the state until we see the same state twice in a row.
+	 */
+
+	read_state(&state2);
+	do {
+		state = state2;
+		read_state(&state2);
+	} while (*pun != *pun2);
+
+	/*
+	 * Normalize the state; 12hr -> 24hr, BCD -> decimal.
+	 * The order is a bit awkward because we need to interpret
+	 * the HOUR_PM flag before we adjust for BCD.
+	 */
+
+	if (state.status_b & STATUS_B_24HR) {
+		pm = false;
+	} else {
+		pm = ((state.hour & HOUR_PM) == HOUR_PM);
+		state.hour &= ~HOUR_PM;
+	}
+
+	if (!(state.status_b & STATUS_B_BIN)) {
+		u8_t *cp = (u8_t *) &state;
+		int i;
+
+		for (i = 0; i < NR_BCD_VALS; ++i) {
+			*cp = decode_bcd(*cp);
+			++cp;
+		}
+	}
+
+	if (pm) {
+		state.hour = (state.hour + 12) % 24;
+	}
+
+	/*
+	 * Convert date/time to epoch time. We don't care about
+	 * timezones here, because we're just creating a mapping
+	 * that results in a monotonic clock; the absolute value
+	 * is irrelevant.
+	 */
+
+	epoch = hinnant(state.year + 2000, state.month, state.day);
+	epoch *= 86400; /* seconds per day */
+	epoch += state.hour * 3600; /* seconds per hour */
+	epoch += state.minute * 60; /* seconds per minute */
+	epoch += state.second;
+
+	return epoch;
+}
+
+static int init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+static const struct counter_driver_api api = {
+	.read = read
+};
+
```

Comment:
would be good to add `counter_config_info` to the device then at least frequency and max_top_value and direction can be determined through counter API.

Reply:
@nordic-krch Good idea. It's a great project for a summer intern... so I've assigned one. :)